vyper/builtin_functions/functions.py

```diff
@@ -461,37 +500,34 @@ class Concat(_BuiltinFunction):
     _id = "concat"
 
     def fetch_call_return(self, node):
-        if len(node.args) < 2:
-            raise ArgumentException("Invalid argument count: expected at least 2", node)
-
-        if node.keywords:
-            raise ArgumentException("Keyword arguments are not accepted here", node.keywords[0])
-
-        type_ = None
-        for expected in (BytesAbstractType(), StringPrimitive()):
-            try:
-                validate_expected_type(node.args[0], expected)
-                type_ = expected
-            except (InvalidType, TypeMismatch):
-                pass
-        if type_ is None:
-            raise TypeMismatch("Concat values must be bytes or string", node.args[0])
-
-        length = 0
-        for arg in node.args[1:]:
-            validate_expected_type(arg, type_)
+        arg_types = self.infer_arg_types(node)
 
         length = 0
-        for arg in node.args:
-            length += get_possible_types_from_node(arg).pop().length
+        for arg_t in arg_types:
+            length += arg_t.length
 
-        if isinstance(type_, BytesAbstractType):
+        if isinstance(arg_types[0], BytesAbstractType):
             return_type = BytesArrayDefinition()
         else:
             return_type = StringDefinition()
         return_type.set_length(length)
         return return_type
 
+    def infer_arg_types(self, node):
+        if len(node.args) < 2:
+            raise ArgumentException("Invalid argument count: expected at least 2", node)
+
+        if node.keywords:
+            raise ArgumentException("Keyword arguments are not accepted here", node.keywords[0])
+
+        ret = []
+        for arg in node.args:
+            validate_expected_type(arg, (BytesAbstractType(), StringDefinition()))
```

Comment:
this is probably a good place to check that the args are all either Bytes or Strings (which is currently done in `build_IR`).